File: vnc.go
package main

import (
    "bufio"
    "fmt"
    "net"
    "log"
    "os/exec"
    "time"
    "os"
    "github.com/bclicn/color"
)

const (
  v0 = "RFB 000.000\n"  // UltraVNC repeater
  v1 = "RFB 003.003\n"
  v2 = "RFB 003.006\n"  // UltraVNC
  v3 = "RFB 003.889\n"  // Apple Remote Desktop
  v4 = "RFB 003.007\n"
  v5 = "RFB 003.008\n"
  v6 = "RFB 004.000\n"  // Intel AMT KVM
  v7 = "RFB 004.001\n"  // RealVNC 4.6
  TCP_PORT = "5900"  // TCP PORT
  TIMEOUT = 2  // timeout x second
)

func system(cmd string, arg ...string) string{
    out, err := exec.Command(cmd, arg...).Output()
    if err != nil {
        log.Fatal(err)
    }
    return string(out)
}
func checkSecurityType(ip string, port string) bool{
  conn, err := net.DialTimeout("tcp", ip + ":" + port, TIMEOUT * time.Second)
  if err != nil {
    return false
  }
  connbuf := bufio.NewReader(conn)
  str, err := connbuf.ReadString('\n')
  if len(str)>0 {
    if v5 == str {
      // vncsnapshot https://github.com/shamun/vncsnapshot
      // system("timeout","60","vncsnapshot","-allowblank","-vncQuality","6","-quality","60",ip,""+ip+".jpg")
      fmt.Println(color.Green("No Authentication: ") + color.LightYellow(ip))
      return true
    }
  }
  if err!= nil {
      return false
  }
  return false
}

func main() {
    scanner := bufio.NewScanner(os.Stdin)
    for {
        fmt.Print(color.BYellow("Please enter full file path to list of IPs.") + color.LightRed("\nExample: ") + "/home/user/Documents/list.txt\nFull path: ")
        scanner.Scan()
        file_path := scanner.Text()
        if len(file_path) != 0 {
            fmt.Print("\nRunning ...\n")
            file, err := os.Open(file_path)
            if err != nil {
                fmt.Print(color.LightRed("\nPlease enter full file path\n"))
                break
            }
            defer file.Close()
            scanner := bufio.NewScanner(file)
            for scanner.Scan() {
                checkSecurityType(scanner.Text(),TCP_PORT)
            }
            if err := scanner.Err(); err != nil {
                log.Fatal(err)
                break
            }
        } else {
            break
        }
        fmt.Println("\nProgram complete, enjoy sifting through your results :-)")
        break
    }
}

File: vnc.py
import socket
import os

TCP_PORT = 5900
CURRENT_INDEX = 0
TIMEOUT = 3

def checkSecurityType(TCP_IP):
    #types 0 Invalid | 1 NONE | 2 VNC Authentication
    authentication = 0
    vnc = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
    vnc.settimeout(TIMEOUT)
    try:
        vnc.connect( ( TCP_IP, int( TCP_PORT ) ) )
        vnc_ver = vnc.recv(12)
        print(vnc_ver)
        if 'RFB' not in str(vnc_ver):
            return authentication
        vnc.send(vnc_ver)
        types = ord( vnc.recv(1) )
        if not types:
            return authentication
        # print("security types: " + str( types ))
        for i in range(0,types):
            if ord( vnc.recv(1) ) == 1:
                authentication = 1
            else:
                pass
        vnc.shutdown(socket.SHUT_WR)
        vnc.close()
    except socket.error:
        vnc.close()
        pass
    except socket.timeout:
        vnc.close()
        pass
    return authentication

if __name__ == '__main__':
    try: input = raw_input
    except NameError: pass
    file_path = input("Please enter full file path to list of IPs.\nExample: /home/user/Documents/list.txt\nFull path: ")
    print("Note ... current index will be kept in new file 'vnc-log.txt' if for some reason the program is interrupted")
    print("\nRunning ...")
    with open(file_path, 'r') as file:
        for line in file:
            line_ip = line.strip('\n')
            vncType = checkSecurityType(line_ip)
            CURRENT_INDEX = CURRENT_INDEX + 1
            os.system("echo " + str(CURRENT_INDEX) + " > vnc-log.txt")
            if vncType == 1:
                command = "timeout 60 vncsnapshot -allowblank -vncQuality 6 -quality 60 " + line_ip + " " + line_ip + ".jpg> /dev/null 2>&1"
                os.system(command)
                print("done: " + line_ip)
            else:
                pass
    print("Program complete, enjoy sifting through your results :-)")
